Dev/link parser

Valpydeveloper opened the pull request:
1) Add bs4 linkedin parser

Valpydeveloper committed:
Add linkedin pars

Valpydeveloper committed:
add linkedin parser

saigono commented on backend/parsing/link_vac_bs4.py:
```diff
@@ -0,0 +1,123 @@
+
+
+
+parse_all_page(0)
```
вызов не нужен (пока что)

Valpydeveloper replied:
убрал

saigono commented on backend/parsing/link_vac_bs4.py:
```diff
@@ -0,0 +1,123 @@
+    for vacancy in all_vacancy_on_page:
+        parse_vacancy(vacancy)
+
+def parse_all_page(start):
```
Не очень понятно зачем нужно передавать этот параметр в функцию

Valpydeveloper replied:
тоже убрал )

saigono reviewed (changes requested):
Два главных комментария:
- убрать скрапи
- возвращать значение из функции "parse_all_page" вместо того чтобы писать в json файл

Valpydeveloper committed:
1) added a inkeding parser to beautifuloup 2) removed all files related to the inkeding parser from scrapy 3)moved the file link_vac_bs4.py to the parsing folder

Valpydeveloper commented on backend/parsing/link_vac_bs4.py:
```diff
@@ -0,0 +1,123 @@
+import json
+import time
+
+start = -25
```
Переменная была в изначальном варианте, когда не знал как задать цикл ) убрал

Valpydeveloper commented on backend/parsing/link_vac_bs4.py:
```diff
@@ -0,0 +1,123 @@
+    ten_chars_before_refid = match_before_refid.group(0)
+    tracking_id = match_trackingid_position.group(1)
+    detail_url = f"https://www.linkedin.com/jobs-guest/jobs/api/jobPosting/{ten_chars_before_refid}?trackingId={tracking_id}"
+    print(detail_url)
```
print(detail_url) был нужен для отладки, тоже убрал. Строки 63,64,65 нужны для получения ссылки на страницу с описанием вакансии

Valpydeveloper commented on backend/parsing/link_vac_bs4.py:
```diff
@@ -0,0 +1,123 @@
+    company_name = vacancy.find('h4', class_='base-search-card__subtitle').text
+    title = vacancy.find('h3', class_='base-search-card__title').text
+    location = vacancy.find('span', class_='job-search-card__location').text
+    url_vacancy = vacancy.find("a", class_="base-card__full-link absolute top-0 right-0 bottom-0 left-0 p-0 z-[2]").get("href")
```
Сделал все кавычки в файле одинарными

Valpydeveloper commented on backend/parsing/link_vac_bs4.py:
```diff
@@ -0,0 +1,123 @@
+from bs4 import BeautifulSoup
```
Отсортировал через isort

Valpydeveloper commented on backend/parsing/link_vac_bs4.py:
```diff
@@ -0,0 +1,123 @@
+        soup = BeautifulSoup(r.text, "lxml")
+        all_vacancy_on_page = soup.findAll("div", class_="base-card relative w-full hover:no-underline focus:no-underline base-card--link base-search-card base-search-card--link job-search-card")
+        parse_one_page(all_vacancy_on_page)
+        start += 25
```
Единственный комментарий, который пока не понял ) можно поподробнее описать суть проблемы ? Одна страница линки по тому урлу на который делается первый запрос, возвращает 25 объектов класса "base-card relative w-full hover:no-underline focus:no-underline base-card--link base-search-card base-search-card--link job-search-card", для каждого из них выполняется функция которая парсит эту самую одну страницу

saigono replied:
Непонятно что за магическое число 25, лучше написать небольшой комментарий

Valpydeveloper replied:
Тут я ошибся, изменил значение с 25 на 10. В самом коде добавил документацию, где упомянул для чего это значение

Valpydeveloper commented on backend/parsing/link_vac_bs4.py:
```diff
@@ -0,0 +1,123 @@
+        time.sleep(1)
+    start = 0
+    print("Начинаю поиск по Беларуси")
+    while start < 1000:
```
Ну да, разница только в них, суть двух циклов в том, чтобы после первого цикла значение старт обнулялось. Попробую придумать как укоротить, можно сделать надцикл for, который будет идти по списку geoid и для каждого из них вызывать цикл for. Я правильно понял замечание ?

saigono replied:
Да, правильно

Valpydeveloper replied:
Сделано

Valpydeveloper committed:
1) moved the variable start inside the function parse_all_page, 2) removed print(detail_url) from the function parse_vacancy, 3) made all quotes in the file single quotes, 4) removed the call to the function parse_all_page

saigono commented on backend/parsing/link_vac_bs4.py:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    with open("vacancy_lit.json", "a", encoding="utf-8") as json_file:
+        json.dump(vacancy, json_file, indent=4, ensure_ascii=False)
```
Всё ещё нужно поменять :)

saigono replied:
Жду…

Valpydeveloper committed:
1) Implemented a for loop over geoid for the parse_all_page function and removed the unnecessary while loop. 2) Corrected the change of the start variable value from 25 to 10. 3) Added a function for handling the 429 error. 4) Added documentation in Russian.

Valpydeveloper commented on backend/parsing/link_vac_bs4.py:
```diff
@@ -0,0 +1,123 @@
+    start = 0
+    print("Начинаю поиск по Беларуси")
+    while start < 1000:
+        url = "https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search?keywords=&location=&geoId=101705918&trk=public_jobs_jobs-search-bar_search-submit&start=" + str(start)
```
Переделал в f строку

Valpydeveloper committed:
fix: added a status check for the detail_url

Valpydeveloper committed:
1) Removed the JSON file writing. Now all vacancies are added to a list.

2) Added a script that interrupts the loops when the parser reaches empty pages.

3) Updated the documentation.
